// sigstore/dist/tlog/verify/body.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyTLogBody = void 0;
/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const error_1 = require("../../error");
const util_1 = require("../../util");
const TLOG_MISMATCH_ERROR_MSG = 'bundle content and tlog entry do not match';
// Compare the given tlog entry to the given bundle
function verifyTLogBody(entry, bundleContent) {
    const { kind, version } = entry.kindVersion;
    const body = JSON.parse(entry.canonicalizedBody.toString('utf8'));
    try {
        if (kind !== body.kind || version !== body.apiVersion) {
            throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
        }
        switch (body.kind) {
            case 'dsse':
                verifyDSSETLogBody(body, bundleContent);
                break;
            case 'intoto':
                verifyIntotoTLogBody(body, bundleContent);
                break;
            case 'hashedrekord':
                verifyHashedRekordTLogBody(body, bundleContent);
                break;
            default:
                throw new error_1.VerificationError(`unsupported kind in tlog entry: ${kind}`);
        }
        return true;
    }
    catch (e) {
        return false;
    }
}
exports.verifyTLogBody = verifyTLogBody;
// Compare the given intoto tlog entry to the given bundle
function verifyDSSETLogBody(tlogEntry, content) {
    if (content?.$case !== 'dsseEnvelope') {
        throw new error_1.VerificationError(`unsupported bundle content: ${content?.$case || 'unknown'}`);
    }
    const dsse = content.dsseEnvelope;
    switch (tlogEntry.apiVersion) {
        case '0.0.1':
            verifyDSSE001TLogBody(tlogEntry, dsse);
            break;
        default:
            throw new error_1.VerificationError(`unsupported dsse version: ${tlogEntry.apiVersion}`);
    }
}
// Compare the given intoto tlog entry to the given bundle
function verifyIntotoTLogBody(tlogEntry, content) {
    if (content?.$case !== 'dsseEnvelope') {
        throw new error_1.VerificationError(`unsupported bundle content: ${content?.$case || 'unknown'}`);
    }
    const dsse = content.dsseEnvelope;
    switch (tlogEntry.apiVersion) {
        case '0.0.2':
            verifyIntoto002TLogBody(tlogEntry, dsse);
            break;
        default:
            throw new error_1.VerificationError(`unsupported intoto version: ${tlogEntry.apiVersion}`);
    }
}
// Compare the given hashedrekord tlog entry to the given bundle
function verifyHashedRekordTLogBody(tlogEntry, content) {
    if (content?.$case !== 'messageSignature') {
        throw new error_1.VerificationError(`unsupported bundle content: ${content?.$case || 'unknown'}`);
    }
    const messageSignature = content.messageSignature;
    switch (tlogEntry.apiVersion) {
        case '0.0.1':
            verifyHashedrekor001TLogBody(tlogEntry, messageSignature);
            break;
        default:
            throw new error_1.VerificationError(`unsupported hashedrekord version: ${tlogEntry.apiVersion}`);
    }
}
// Compare the given dsse v0.0.1 tlog entry to the given DSSE envelope.
function verifyDSSE001TLogBody(tlogEntry, dsse) {
    // Collect all of the signatures from the DSSE envelope
    // Turns them into base64-encoded strings for comparison
    const dsseSigs = dsse.signatures.map((signature) => signature.sig.toString('base64'));
    // Collect all of the signatures from the tlog entry
    const tlogSigs = tlogEntry.spec.signatures?.map((signature) => signature.signature);
    // Ensure the bundle's DSSE and the tlog entry contain the same number of signatures
    if (dsseSigs.length !== tlogSigs?.length) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
    // Ensure that every signature in the bundle's DSSE is present in the tlog entry
    if (!dsseSigs.every((dsseSig) => tlogSigs.includes(dsseSig))) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
    // Ensure the digest of the bundle's DSSE payload matches the digest in the
    // tlog entry
    const dssePayloadHash = util_1.crypto.hash(dsse.payload).toString('hex');
    if (dssePayloadHash !== tlogEntry.spec.payloadHash?.value) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
}
// Compare the given intoto v0.0.2 tlog entry to the given DSSE envelope.
function verifyIntoto002TLogBody(tlogEntry, dsse) {
    // Collect all of the signatures from the DSSE envelope
    // Turns them into base64-encoded strings for comparison
    const dsseSigs = dsse.signatures.map((signature) => signature.sig.toString('base64'));
    // Collect all of the signatures from the tlog entry
    // Remember that tlog signastures are double base64-encoded
    const tlogSigs = tlogEntry.spec.content.envelope?.signatures.map((signature) => (signature.sig ? util_1.encoding.base64Decode(signature.sig) : ''));
    // Ensure the bundle's DSSE and the tlog entry contain the same number of signatures
    if (dsseSigs.length !== tlogSigs?.length) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
    // Ensure that every signature in the bundle's DSSE is present in the tlog entry
    if (!dsseSigs.every((dsseSig) => tlogSigs.includes(dsseSig))) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
    // Ensure the digest of the bundle's DSSE payload matches the digest in the
    // tlog entry
    const dssePayloadHash = util_1.crypto.hash(dsse.payload).toString('hex');
    if (dssePayloadHash !== tlogEntry.spec.content.payloadHash?.value) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
}
// Compare the given hashedrekord v0.0.1 tlog entry to the given message
// signature
function verifyHashedrekor001TLogBody(tlogEntry, messageSignature) {
    // Ensure that the bundles message signature matches the tlog entry
    const msgSig = messageSignature.signature.toString('base64');
    const tlogSig = tlogEntry.spec.signature.content;
    if (msgSig !== tlogSig) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
    // Ensure that the bundle's message digest matches the tlog entry
    const msgDigest = messageSignature.messageDigest?.digest.toString('hex');
    const tlogDigest = tlogEntry.spec.data.hash?.value;
    if (msgDigest !== tlogDigest) {
        throw new error_1.VerificationError(TLOG_MISMATCH_ERROR_MSG);
    }
}

// SIG // Begin signature block
// SIG // MIIoQAYJKoZIhvcNAQcCoIIoMTCCKC0CAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // HByCaDPlmvZBHmkC0mGLUsXZNiRp0pVM9XA5pOyP/6mg
// SIG // gg2LMIIGCTCCA/GgAwIBAgITMwAAA4TZaH1mzHVLoQAA
// SIG // AAADhDANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTIzMDcxMzIzNDUzOFoX
// SIG // DTI0MDkxNTIzNDUzOFowgYgxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xMjAwBgNVBAMTKU1pY3Jvc29mdCAzcmQgUGFydHkg
// SIG // QXBwbGljYXRpb24gQ29tcG9uZW50MIIBIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAQ8AMIIBCgKCAQEAznLqoz/10YOqShXj
// SIG // Gln6TE4BGUwZ3uJsXV3BIhK2/5WfWH8YMWvFoTgOul1P
// SIG // rrDlA5qOs/1qSNylRSLn+DKx4e+sQC/x8qVPWx+4bwhu
// SIG // NHaNGEMZsppPjtuc10OmdECsioQzVOyu/TlVohP0DUT1
// SIG // w8V4+uR06k9qm+XVcw84nrH6xac865B7TswYpBBBZdaG
// SIG // X5mTst5UnoomlS8JIGqD6WDAyvlmwDRl0C0Omj+b8lg7
// SIG // ZJ4wz9hfYI6eo/kOu2wgYyLPBaXR9iDehwLxN1/xOGDi
// SIG // Dq2WNiWFM/aaBYda645BrjOAk9rc3aoX5AdvEDp7oRMi
// SIG // fgJgdwdNaC/WoTUomQIDAQABo4IBczCCAW8wHwYDVR0l
// SIG // BBgwFgYKKwYBBAGCN0wRAQYIKwYBBQUHAwMwHQYDVR0O
// SIG // BBYEFBzcFkQMR/V8OGV/cF0ZRXO9ery8MEUGA1UdEQQ+
// SIG // MDykOjA4MR4wHAYDVQQLExVNaWNyb3NvZnQgQ29ycG9y
// SIG // YXRpb24xFjAUBgNVBAUTDTIzMTUyMis1MDExNDgwHwYD
// SIG // VR0jBBgwFoAUSG5k5VAF04KqFzc3IrVtqMp1ApUwVAYD
// SIG // VR0fBE0wSzBJoEegRYZDaHR0cDovL3d3dy5taWNyb3Nv
// SIG // ZnQuY29tL3BraW9wcy9jcmwvTWljQ29kU2lnUENBMjAx
// SIG // MV8yMDExLTA3LTA4LmNybDBhBggrBgEFBQcBAQRVMFMw
// SIG // UQYIKwYBBQUHMAKGRWh0dHA6Ly93d3cubWljcm9zb2Z0
// SIG // LmNvbS9wa2lvcHMvY2VydHMvTWljQ29kU2lnUENBMjAx
// SIG // MV8yMDExLTA3LTA4LmNydDAMBgNVHRMBAf8EAjAAMA0G
// SIG // CSqGSIb3DQEBCwUAA4ICAQAnvqLcbzrt8+kzifeRGz6b
// SIG // UaHMp1aMqJtfMarjiUinqgzSLbtqZslaWomQ4T7LS8rq
// SIG // NJl8jSz4d0wJfm6LblH6RWQzhs4JqllQa3R0MbLt7fvs
// SIG // ehsNASX0L5/3NP5VP40U0UwKe0u4wKo6qcjHNMr66y+s
// SIG // b1DZ965C5Hctslwx1qR3RMeSXaSQ6YJ+Y9bb9xhnbXmH
// SIG // YWZW9h6JWnFYriUm3jHRubAw7KXMPMWNQ7OBTF6uzGKi
// SIG // 2ELK/CsFyWfeUm6GdPVf+TDsKShhoZXtAXOZwBK3SQTY
// SIG // 42gSwTBg4oPTldGd2SH2OIzhZXkxLCuSzk0LUmjVQ+K2
// SIG // xvkJzLMHi2oO6lTO8+SHjqhO1GPwnwhJLJfoaynNzAb3
// SIG // XMj3CdvPcd4EZDh9PRSI0SAHxaQQ0pD4+B1uSyqjUr9w
// SIG // GxZbisemYGzWIcB8UEsQ6mLE5QI7Z2zI2CNkU60iDI4B
// SIG // mEb38Gue8qcj3Zcs5teEOVDYILFsq9ToPkG0HluvkQhu
// SIG // hblf5xAGKzIHh5TEAXedGlYU3QXquQkdxs+quGlS+5xH
// SIG // Q0enmLi9FNXuBKC9jmAaJN8uKwS96+e5yxBGesTBgHuX
// SIG // hwT7YHLj8OUv2iQ66KXrvuAV0ouUf6hpLU5Gyn5STAJX
// SIG // 1xy+muuKPPIfIvGWeBxxzONasTAedo/W4GggkBSRZSyT
// SIG // 7zCCB3owggVioAMCAQICCmEOkNIAAAAAAAMwDQYJKoZI
// SIG // hvcNAQELBQAwgYgxCzAJBgNVBAYTAlVTMRMwEQYDVQQI
// SIG // EwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4w
// SIG // HAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xMjAw
// SIG // BgNVBAMTKU1pY3Jvc29mdCBSb290IENlcnRpZmljYXRl
// SIG // IEF1dGhvcml0eSAyMDExMB4XDTExMDcwODIwNTkwOVoX
// SIG // DTI2MDcwODIxMDkwOVowfjELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEoMCYGA1UEAxMfTWljcm9zb2Z0IENvZGUgU2lnbmlu
// SIG // ZyBQQ0EgMjAxMTCCAiIwDQYJKoZIhvcNAQEBBQADggIP
// SIG // ADCCAgoCggIBAKvw+nIQHC6t2G6qghBNNLrytlghn0Ib
// SIG // KmvpWlCquAY4GgRJun/DDB7dN2vGEtgL8DjCmQawyDnV
// SIG // ARQxQtOJDXlkh36UYCRsr55JnOloXtLfm1OyCizDr9mp
// SIG // K656Ca/XllnKYBoF6WZ26DJSJhIv56sIUM+zRLdd2MQu
// SIG // A3WraPPLbfM6XKEW9Ea64DhkrG5kNXimoGMPLdNAk/jj
// SIG // 3gcN1Vx5pUkp5w2+oBN3vpQ97/vjK1oQH01WKKJ6cuAS
// SIG // OrdJXtjt7UORg9l7snuGG9k+sYxd6IlPhBryoS9Z5JA7
// SIG // La4zWMW3Pv4y07MDPbGyr5I4ftKdgCz1TlaRITUlwzlu
// SIG // ZH9TupwPrRkjhMv0ugOGjfdf8NBSv4yUh7zAIXQlXxgo
// SIG // tswnKDglmDlKNs98sZKuHCOnqWbsYR9q4ShJnV+I4iVd
// SIG // 0yFLPlLEtVc/JAPw0XpbL9Uj43BdD1FGd7P4AOG8rAKC
// SIG // X9vAFbO9G9RVS+c5oQ/pI0m8GLhEfEXkwcNyeuBy5yTf
// SIG // v0aZxe/CHFfbg43sTUkwp6uO3+xbn6/83bBm4sGXgXvt
// SIG // 1u1L50kppxMopqd9Z4DmimJ4X7IvhNdXnFy/dygo8e1t
// SIG // wyiPLI9AN0/B4YVEicQJTMXUpUMvdJX3bvh4IFgsE11g
// SIG // lZo+TzOE2rCIF96eTvSWsLxGoGyY0uDWiIwLAgMBAAGj
// SIG // ggHtMIIB6TAQBgkrBgEEAYI3FQEEAwIBADAdBgNVHQ4E
// SIG // FgQUSG5k5VAF04KqFzc3IrVtqMp1ApUwGQYJKwYBBAGC
// SIG // NxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8G
// SIG // A1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAUci06AjGQ
// SIG // Q7kUBU7h6qfHMdEjiTQwWgYDVR0fBFMwUTBPoE2gS4ZJ
// SIG // aHR0cDovL2NybC5taWNyb3NvZnQuY29tL3BraS9jcmwv
// SIG // cHJvZHVjdHMvTWljUm9vQ2VyQXV0MjAxMV8yMDExXzAz
// SIG // XzIyLmNybDBeBggrBgEFBQcBAQRSMFAwTgYIKwYBBQUH
// SIG // MAKGQmh0dHA6Ly93d3cubWljcm9zb2Z0LmNvbS9wa2kv
// SIG // Y2VydHMvTWljUm9vQ2VyQXV0MjAxMV8yMDExXzAzXzIy
// SIG // LmNydDCBnwYDVR0gBIGXMIGUMIGRBgkrBgEEAYI3LgMw
// SIG // gYMwPwYIKwYBBQUHAgEWM2h0dHA6Ly93d3cubWljcm9z
// SIG // b2Z0LmNvbS9wa2lvcHMvZG9jcy9wcmltYXJ5Y3BzLmh0
// SIG // bTBABggrBgEFBQcCAjA0HjIgHQBMAGUAZwBhAGwAXwBw
// SIG // AG8AbABpAGMAeQBfAHMAdABhAHQAZQBtAGUAbgB0AC4g
// SIG // HTANBgkqhkiG9w0BAQsFAAOCAgEAZ/KGpZjgVHkaLtPY
// SIG // dGcimwuWEeFjkplCln3SeQyQwWVfLiw++MNy0W2D/r4/
// SIG // 6ArKO79HqaPzadtjvyI1pZddZYSQfYtGUFXYDJJ80hpL
// SIG // HPM8QotS0LD9a+M+By4pm+Y9G6XUtR13lDni6WTJRD14
// SIG // eiPzE32mkHSDjfTLJgJGKsKKELukqQUMm+1o+mgulaAq
// SIG // PyprWEljHwlpblqYluSD9MCP80Yr3vw70L01724lruWv
// SIG // J+3Q3fMOr5kol5hNDj0L8giJ1h/DMhji8MUtzluetEk5
// SIG // CsYKwsatruWy2dsViFFFWDgycScaf7H0J/jeLDogaZiy
// SIG // WYlobm+nt3TDQAUGpgEqKD6CPxNNZgvAs0314Y9/HG8V
// SIG // fUWnduVAKmWjw11SYobDHWM2l4bf2vP48hahmifhzaWX
// SIG // 0O5dY0HjWwechz4GdwbRBrF1HxS+YWG18NzGGwS+30HH
// SIG // Diju3mUv7Jf2oVyW2ADWoUa9WfOXpQlLSBCZgB/QACnF
// SIG // sZulP0V3HjXG0qKin3p6IvpIlR+r+0cjgPWe+L9rt0uX
// SIG // 4ut1eBrs6jeZeRhL/9azI2h15q/6/IvrC4DqaTuv/DDt
// SIG // BEyO3991bWORPdGdVk5Pv4BXIqF4ETIheu9BCrE/+6jM
// SIG // pF3BoYibV3FWTkhFwELJm3ZbCoBIa/15n8G9bW1qyVJz
// SIG // Ew16UM0xghoNMIIaCQIBATCBlTB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExAhMzAAADhNlofWbMdUuhAAAA
// SIG // AAOEMA0GCWCGSAFlAwQCAQUAoIGuMBkGCSqGSIb3DQEJ
// SIG // AzEMBgorBgEEAYI3AgEEMBwGCisGAQQBgjcCAQsxDjAM
// SIG // BgorBgEEAYI3AgEVMC8GCSqGSIb3DQEJBDEiBCC87lEb
// SIG // C5GBqXJnLNHSecSFvZ9XKz3bxqmD9dNF33oD+DBCBgor
// SIG // BgEEAYI3AgEMMTQwMqAUgBIATQBpAGMAcgBvAHMAbwBm
// SIG // AHShGoAYaHR0cDovL3d3dy5taWNyb3NvZnQuY29tMA0G
// SIG // CSqGSIb3DQEBAQUABIIBAAcDDLKoJ7WxnjzrKsMbiGBY
// SIG // ymqKwdTeISl7b3XENjgsZFjgnBAUMAJgMge6jjX3yIf5
// SIG // XLe2m/im4V8D9hl53j8w7sL+4gVHfe6ZkTjQgtD+ZQ8g
// SIG // V1u11/qjyARU0oYzivd07DKp6GlFIj91rFCohzq74SU3
// SIG // HM/gLdx+/k+igtWysGmsL090n2mloQYvD4/qZ9c3d0En
// SIG // 9tFpV0Lqac6vvrlMRlUXniSOEksC1U+Ib1Ky9t9zHnB8
// SIG // SzOXmVu8DECyJIyFS3NRDmwMD5xDvkl9DrCO5QNZP1yY
// SIG // 9424sw+2Gt/CEwWTlq0BY+wvKsS9Rz2J4kfwvO2JuM/V
// SIG // Uhj6/kL3GNOhgheXMIIXkwYKKwYBBAGCNwMDATGCF4Mw
// SIG // ghd/BgkqhkiG9w0BBwKgghdwMIIXbAIBAzEPMA0GCWCG
// SIG // SAFlAwQCAQUAMIIBUgYLKoZIhvcNAQkQAQSgggFBBIIB
// SIG // PTCCATkCAQEGCisGAQQBhFkKAwEwMTANBglghkgBZQME
// SIG // AgEFAAQgy1C8pR9LKy+361Iiw3PK8tzYGfl3GB31JKUK
// SIG // bewlHdsCBmUokN2ZOBgTMjAyMzExMTUwMDE0MTYuODQ3
// SIG // WjAEgAIB9KCB0aSBzjCByzELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjElMCMGA1UECxMcTWljcm9zb2Z0IEFtZXJpY2EgT3Bl
// SIG // cmF0aW9uczEnMCUGA1UECxMeblNoaWVsZCBUU1MgRVNO
// SIG // OkRDMDAtMDVFMC1EOTQ3MSUwIwYDVQQDExxNaWNyb3Nv
// SIG // ZnQgVGltZS1TdGFtcCBTZXJ2aWNloIIR7TCCByAwggUI
// SIG // oAMCAQICEzMAAAHSISQxSli/LREAAQAAAdIwDQYJKoZI
// SIG // hvcNAQELBQAwfDELMAkGA1UEBhMCVVMxEzARBgNVBAgT
// SIG // Cldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAc
// SIG // BgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQG
// SIG // A1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIw
// SIG // MTAwHhcNMjMwNTI1MTkxMjIxWhcNMjQwMjAxMTkxMjIx
// SIG // WjCByzELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hp
// SIG // bmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoT
// SIG // FU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjElMCMGA1UECxMc
// SIG // TWljcm9zb2Z0IEFtZXJpY2EgT3BlcmF0aW9uczEnMCUG
// SIG // A1UECxMeblNoaWVsZCBUU1MgRVNOOkRDMDAtMDVFMC1E
// SIG // OTQ3MSUwIwYDVQQDExxNaWNyb3NvZnQgVGltZS1TdGFt
// SIG // cCBTZXJ2aWNlMIICIjANBgkqhkiG9w0BAQEFAAOCAg8A
// SIG // MIICCgKCAgEA3GCIQtECP0j2k+fp2EgbEnYQTztklzON
// SIG // FcsoPBatUz6zDQ88QocVlA+32ysGBkrDjYLY41XneR0o
// SIG // eYpe+YzWnG+k62cLl/kNvvn9v/AgsrvE2NK8RCJw904+
// SIG // dm1s+UEVlUHo33QgP5+mc4p4XGWe/lXpVoE32Au6w64V
// SIG // 90wV/PXpal5iEjlSFnl4kSvIfYR+kl4I4UBhkG39m0jO
// SIG // 7hpCsyf6Wdi75U86vpjlvYj9o3GBAOG6T3rq6YJS1VEp
// SIG // xhkC5o0Ke0nctcBOdlozynS5wReiW5kkRDPw4/We6JMU
// SIG // TVI2EnYMr2dS1kopYVf8HoGF1UOj/TCo0ZxPAiOSjB5q
// SIG // uxLB4SAStAffXRUMnS9b3bgu3lhI88+eWfw8h+h5dkjz
// SIG // n+SmeJR3+jFMLYGn/jHmcZFtIsYf7y9Zl/FbCUx2vAB/
// SIG // tJf/XZ1sxTBRSRRu4PeZxyejJmD635qHijkaMhEEbErr
// SIG // Nqgq6TI4Pz92fViQLRRqF2DA9fTa8xuFBT6VW5dzdV7x
// SIG // uP7pXaNfmpwaQ2IcUFO5NYHx5+kMXVFNE7KyfyQry330
// SIG // vPW+FV5SGsZ1RekGYfie8S1DRrRJo+ncD86Any+86g2S
// SIG // b5rL2DZNZXG8lcTCIAxT6ANn2T1xxe3cWoTNmRuwALd7
// SIG // 2RIjLQuhP3Ii0XCMM5SawYhxHpyD78VteasCAwEAAaOC
// SIG // AUkwggFFMB0GA1UdDgQWBBQgtp+3J0d06D/5GxGdlyRR
// SIG // ZmdZJDAfBgNVHSMEGDAWgBSfpxVdAF5iXYP05dJlpxtT
// SIG // NRnpcjBfBgNVHR8EWDBWMFSgUqBQhk5odHRwOi8vd3d3
// SIG // Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2NybC9NaWNyb3Nv
// SIG // ZnQlMjBUaW1lLVN0YW1wJTIwUENBJTIwMjAxMCgxKS5j
// SIG // cmwwbAYIKwYBBQUHAQEEYDBeMFwGCCsGAQUFBzAChlBo
// SIG // dHRwOi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2Nl
// SIG // cnRzL01pY3Jvc29mdCUyMFRpbWUtU3RhbXAlMjBQQ0El
// SIG // MjAyMDEwKDEpLmNydDAMBgNVHRMBAf8EAjAAMBYGA1Ud
// SIG // JQEB/wQMMAoGCCsGAQUFBwMIMA4GA1UdDwEB/wQEAwIH
// SIG // gDANBgkqhkiG9w0BAQsFAAOCAgEAsOYqtBMtC5QHc9j3
// SIG // KnIUcFuXfn4rTSEgr6g9BOlvwMok5PkzN1oWE2sUNqT+
// SIG // jq+Kjlfg7u/y/pNfIGK9aYbIcVTFZ1zXHY3nDLU1lDlQ
// SIG // MGyaHlh2gIJJzSDM1yt/s4LhuU1XCxIfwlTMtSeyjxPE
// SIG // NAs+ejYFs3fbqxvrIUgypCnYTmD1xfe9DK9RghRaXtaF
// SIG // 4xMumkvh3mr6dsJsuuZm7tAHFDjxJ/oWKoo4C10j6r7h
// SIG // ZjjqOVJ0leug7+RXZKiu6uQ0XqMfUTxGXB1GvtgVCj46
// SIG // fnUNw+HPSB/O9uyozA/Dl2aJg/QPDK/33A5+RSO5X2XP
// SIG // IIWHO7KWCkzv9GfWEUWpoVVscayLATeS61oftfLKFyfv
// SIG // hHko80mHYBlmD4NeoS/+wMnceCs/1gU1JmxoolVHBPyN
// SIG // GkFNlP9Dc1IiBSFvNEr8Xg1p+iPoclYP7HRUgCz+SU+8
// SIG // AHvM077KyNoTKRFS3UrRQDKmpxWKrcGVORaALKD2e4dR
// SIG // JVpii2ZF5PHj7gC87TWNcrHjtE8HbniBnCMCLa5QWINV
// SIG // 6T89fqTIRjFW5Xuih7p2iqOSv5XTjNJkEE+FY4OmW80S
// SIG // FUaA3AK0Of/9PO+YzlLybWMnOx5dOI3Jr4xCFwtx45FG
// SIG // L4BxFi9EBf+H+7ywHIWhGfN/f9xmDnah+VVZ2a1CAUAe
// SIG // g02lCCwwggdxMIIFWaADAgECAhMzAAAAFcXna54Cm0mZ
// SIG // AAAAAAAVMA0GCSqGSIb3DQEBCwUAMIGIMQswCQYDVQQG
// SIG // EwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UE
// SIG // BxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENv
// SIG // cnBvcmF0aW9uMTIwMAYDVQQDEylNaWNyb3NvZnQgUm9v
// SIG // dCBDZXJ0aWZpY2F0ZSBBdXRob3JpdHkgMjAxMDAeFw0y
// SIG // MTA5MzAxODIyMjVaFw0zMDA5MzAxODMyMjVaMHwxCzAJ
// SIG // BgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAw
// SIG // DgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3Nv
// SIG // ZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29m
// SIG // dCBUaW1lLVN0YW1wIFBDQSAyMDEwMIICIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAg8AMIICCgKCAgEA5OGmTOe0ciELeaLL
// SIG // 1yR5vQ7VgtP97pwHB9KpbE51yMo1V/YBf2xK4OK9uT4X
// SIG // YDP/XE/HZveVU3Fa4n5KWv64NmeFRiMMtY0Tz3cywBAY
// SIG // 6GB9alKDRLemjkZrBxTzxXb1hlDcwUTIcVxRMTegCjhu
// SIG // je3XD9gmU3w5YQJ6xKr9cmmvHaus9ja+NSZk2pg7uhp7
// SIG // M62AW36MEBydUv626GIl3GoPz130/o5Tz9bshVZN7928
// SIG // jaTjkY+yOSxRnOlwaQ3KNi1wjjHINSi947SHJMPgyY9+
// SIG // tVSP3PoFVZhtaDuaRr3tpK56KTesy+uDRedGbsoy1cCG
// SIG // MFxPLOJiss254o2I5JasAUq7vnGpF1tnYN74kpEeHT39
// SIG // IM9zfUGaRnXNxF803RKJ1v2lIH1+/NmeRd+2ci/bfV+A
// SIG // utuqfjbsNkz2K26oElHovwUDo9Fzpk03dJQcNIIP8BDy
// SIG // t0cY7afomXw/TNuvXsLz1dhzPUNOwTM5TI4CvEJoLhDq
// SIG // hFFG4tG9ahhaYQFzymeiXtcodgLiMxhy16cg8ML6EgrX
// SIG // Y28MyTZki1ugpoMhXV8wdJGUlNi5UPkLiWHzNgY1GIRH
// SIG // 29wb0f2y1BzFa/ZcUlFdEtsluq9QBXpsxREdcu+N+VLE
// SIG // hReTwDwV2xo3xwgVGD94q0W29R6HXtqPnhZyacaue7e3
// SIG // PmriLq0CAwEAAaOCAd0wggHZMBIGCSsGAQQBgjcVAQQF
// SIG // AgMBAAEwIwYJKwYBBAGCNxUCBBYEFCqnUv5kxJq+gpE8
// SIG // RjUpzxD/LwTuMB0GA1UdDgQWBBSfpxVdAF5iXYP05dJl
// SIG // pxtTNRnpcjBcBgNVHSAEVTBTMFEGDCsGAQQBgjdMg30B
// SIG // ATBBMD8GCCsGAQUFBwIBFjNodHRwOi8vd3d3Lm1pY3Jv
// SIG // c29mdC5jb20vcGtpb3BzL0RvY3MvUmVwb3NpdG9yeS5o
// SIG // dG0wEwYDVR0lBAwwCgYIKwYBBQUHAwgwGQYJKwYBBAGC
// SIG // NxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8G
// SIG // A1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAU1fZWy4/o
// SIG // olxiaNE9lJBb186aGMQwVgYDVR0fBE8wTTBLoEmgR4ZF
// SIG // aHR0cDovL2NybC5taWNyb3NvZnQuY29tL3BraS9jcmwv
// SIG // cHJvZHVjdHMvTWljUm9vQ2VyQXV0XzIwMTAtMDYtMjMu
// SIG // Y3JsMFoGCCsGAQUFBwEBBE4wTDBKBggrBgEFBQcwAoY+
// SIG // aHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3BraS9jZXJ0
// SIG // cy9NaWNSb29DZXJBdXRfMjAxMC0wNi0yMy5jcnQwDQYJ
// SIG // KoZIhvcNAQELBQADggIBAJ1VffwqreEsH2cBMSRb4Z5y
// SIG // S/ypb+pcFLY+TkdkeLEGk5c9MTO1OdfCcTY/2mRsfNB1
// SIG // OW27DzHkwo/7bNGhlBgi7ulmZzpTTd2YurYeeNg2Lpyp
// SIG // glYAA7AFvonoaeC6Ce5732pvvinLbtg/SHUB2RjebYIM
// SIG // 9W0jVOR4U3UkV7ndn/OOPcbzaN9l9qRWqveVtihVJ9Ak
// SIG // vUCgvxm2EhIRXT0n4ECWOKz3+SmJw7wXsFSFQrP8DJ6L
// SIG // GYnn8AtqgcKBGUIZUnWKNsIdw2FzLixre24/LAl4FOmR
// SIG // sqlb30mjdAy87JGA0j3mSj5mO0+7hvoyGtmW9I/2kQH2
// SIG // zsZ0/fZMcm8Qq3UwxTSwethQ/gpY3UA8x1RtnWN0SCyx
// SIG // TkctwRQEcb9k+SS+c23Kjgm9swFXSVRk2XPXfx5bRAGO
// SIG // WhmRaw2fpCjcZxkoJLo4S5pu+yFUa2pFEUep8beuyOiJ
// SIG // Xk+d0tBMdrVXVAmxaQFEfnyhYWxz/gq77EFmPWn9y8FB
// SIG // SX5+k77L+DvktxW/tM4+pTFRhLy/AsGConsXHRWJjXD+
// SIG // 57XQKBqJC4822rpM+Zv/Cuk0+CQ1ZyvgDbjmjJnW4SLq
// SIG // 8CdCPSWU5nR0W2rRnj7tfqAxM328y+l7vzhwRNGQ8cir
// SIG // Ooo6CGJ/2XBjU02N7oJtpQUQwXEGahC0HVUzWLOhcGby
// SIG // oYIDUDCCAjgCAQEwgfmhgdGkgc4wgcsxCzAJBgNVBAYT
// SIG // AlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQH
// SIG // EwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29y
// SIG // cG9yYXRpb24xJTAjBgNVBAsTHE1pY3Jvc29mdCBBbWVy
// SIG // aWNhIE9wZXJhdGlvbnMxJzAlBgNVBAsTHm5TaGllbGQg
// SIG // VFNTIEVTTjpEQzAwLTA1RTAtRDk0NzElMCMGA1UEAxMc
// SIG // TWljcm9zb2Z0IFRpbWUtU3RhbXAgU2VydmljZaIjCgEB
// SIG // MAcGBSsOAwIaAxUAiabSwmbBNOjbZh0MweRdU66BUgag
// SIG // gYMwgYCkfjB8MQswCQYDVQQGEwJVUzETMBEGA1UECBMK
// SIG // V2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwG
// SIG // A1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSYwJAYD
// SIG // VQQDEx1NaWNyb3NvZnQgVGltZS1TdGFtcCBQQ0EgMjAx
// SIG // MDANBgkqhkiG9w0BAQsFAAIFAOj95BwwIhgPMjAyMzEx
// SIG // MTQxMjE4MzZaGA8yMDIzMTExNTEyMTgzNlowdzA9Bgor
// SIG // BgEEAYRZCgQBMS8wLTAKAgUA6P3kHAIBADAKAgEAAgIQ
// SIG // JQIB/zAHAgEAAgITjDAKAgUA6P81nAIBADA2BgorBgEE
// SIG // AYRZCgQCMSgwJjAMBgorBgEEAYRZCgMCoAowCAIBAAID
// SIG // B6EgoQowCAIBAAIDAYagMA0GCSqGSIb3DQEBCwUAA4IB
// SIG // AQCmDfL6wH9PksamWlOQl5wb57mHFaca000ICfIQ7DVo
// SIG // dkDHJlsG9q4Bexo1J9c5pxDfmrnWhAt+Ff8RzO3JkRN3
// SIG // sDgSDefOLudSJaRbdNq4bZfzz9LONa4FLPxSKnQKnPz+
// SIG // mDn3d/7KxXVBBNda6CJyEls8UdSgR+7+eaYOnjod6IPc
// SIG // 7NXbAtDrfseOmrzVLBkLiC3xj2jlRQvpvg7fD2SHriyb
// SIG // 9L0BQRM3Jti4NTaiY75Og0i0s4ho1U4DGyG9JWbfNn3y
// SIG // SfTAoBk/XCBBr+WfQLpGmeFGcgMXfkTo/rq9+obcu4Op
// SIG // nMAIrvcaaumSNj0hU9jcOUx2cuyuYDuo9H/QMYIEDTCC
// SIG // BAkCAQEwgZMwfDELMAkGA1UEBhMCVVMxEzARBgNVBAgT
// SIG // Cldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAc
// SIG // BgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQG
// SIG // A1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIw
// SIG // MTACEzMAAAHSISQxSli/LREAAQAAAdIwDQYJYIZIAWUD
// SIG // BAIBBQCgggFKMBoGCSqGSIb3DQEJAzENBgsqhkiG9w0B
// SIG // CRABBDAvBgkqhkiG9w0BCQQxIgQgoSN2hepJ6kGo808H
// SIG // CSofNlbPl4J/L1dRw8SmEghKKMIwgfoGCyqGSIb3DQEJ
// SIG // EAIvMYHqMIHnMIHkMIG9BCDHgCCT399IvWWzhMVOeexs
// SIG // FjWix9GebOuSRYGtg3mkTjCBmDCBgKR+MHwxCzAJBgNV
// SIG // BAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYD
// SIG // VQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQg
// SIG // Q29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29mdCBU
// SIG // aW1lLVN0YW1wIFBDQSAyMDEwAhMzAAAB0iEkMUpYvy0R
// SIG // AAEAAAHSMCIEIDmPgSCNyW5WCwAh/1Dx+oJJumNYsrFD
// SIG // LumZhc1sNDLKMA0GCSqGSIb3DQEBCwUABIICAAfvey/x
// SIG // XXHwaKXfRT+EojEUJcIb2AU6PwuOAddbdbNlqQZm0Wtc
// SIG // 1p0wCeKKgswN1OGfFIu2E08nun7hsBCgZsgVUlrsg3zV
// SIG // lzB9F5mKL82f28lf5BnIJn2xcXgVtYmpXPITym5ykxcS
// SIG // Tt9VCukokrkLTCLOr6XWJTy6tSeZqW2eXlNK9Zok5hkn
// SIG // 1K7yOkOzc6rYGZt6eMzzth/LQNLdDRaUSsgpvdf2mM0h
// SIG // Io9CV9y/1OBXSYJHcuL6jsEZ/gaXkQFdK+C3dX774mVD
// SIG // BugzKz4JnKb9hSNImKHsld1aHwRAIWPhu7kWJu31frv0
// SIG // NAgvuaoGLL4shja1TS+Kctpi5dlfyEtF6mM09eFGs8Ti
// SIG // 50p1LsPmVpoVtHjrgN65SH7qk0rTUqu1IicB5x9SfYqH
// SIG // QaUaij5LsY4GhfnRm0Xk4AxFI9LU7XmybA+TgWSIuvEy
// SIG // w+G0/OXZlVEJ/DcaB9HPIcEZ45NZ5oX3XFVm+rtg7+ff
// SIG // SRyVserV2ghEev+x4/pnJLaXralp6Xk1lpu09MYBoEIJ
// SIG // ZGTT+T/pR1mny5GkrqpqH/gv8pQzv2PSkJ1A1iwtnxBn
// SIG // 3odNC1Uo6b1K2Rc99j01VKRJnyaaCc68pXfnhs0qjVye
// SIG // XqeGyaHyl+ZIN3tSwPVLlLD/h69DNLfgHUzmoLPgQyic
// SIG // WMS2t79cJeSHGbgs
// SIG // End signature block
